pkg/promtail/targets/filetarget.go

```diff
@@ -198,12 +171,103 @@ func (t *FileTarget) run() {
 			}
 		case err := <-t.watcher.Errors:
 			level.Error(t.logger).Log("msg", "error from fswatch", "error", err)
+		case <-ticker.C:
+			err := t.sync()
+			if err != nil {
+				level.Error(t.logger).Log("msg", "error running sync function", "error", err)
+			}
 		case <-t.quit:
 			return
 		}
 	}
 }
 
+func (t *FileTarget) sync() error {
+
+	// Find list of directories to add to watcher.
+	path, err := filepath.Abs(t.path)
+	if err != nil {
+		return errors.Wrap(err, "filepath.Abs")
+	}
+
+	// Gets current list of files to tail.
+	matches, err := filepath.Glob(path)
+	if err != nil {
+		return errors.Wrap(err, "filepath.Glob")
+	}
+
+	// Get the current unique set of dirs to watch.
+	dirs := map[string]struct{}{}
+	for _, p := range matches {
+		dirs[filepath.Dir(p)] = struct{}{}
+	}
+
+	// If no files exist yet watch the directory specified in the path.
+	if len(matches) == 0 {
+		dirs[filepath.Dir(path)] = struct{}{}
+	}
+
+	// Add any directories which are not already being watched.
+	for dir := range dirs {
+		if _, ok := t.watches[dir]; !ok {
+			level.Debug(t.logger).Log("msg", "sync() watching new directory", "directory", dir)
+			if err := t.watcher.Add(dir); err != nil {
+				level.Error(t.logger).Log("msg", "sync() error adding directory to watcher", "error", err)
+			}
+		}
+	}
+
+	// Remove any directories which no longer need watching.
+	if len(t.watches) > 0 && len(t.watches) != len(dirs) {
```

Comment:
So its possible that a directory is added & removed between a sync, meaning `len(t.watches) == len(dirs)`, which would mean this is skipped.  Remove the if completely, seems like an optimisation.